sys/src/9/ufs/ufs_harvey_vnode.c

```diff
@@ -117,41 +195,40 @@ getnewvnode(MountPoint *mp, vnode **vpp)
 }
 
 void
-releaseufsvnode(vnode *vn)
+releasevnode(vnode *vn)
 {
-	releasevnode(vn);
-}
+	MountPoint *mp = vn->mount;
+	qlock(&mp->vnodes_lock);
+
+	if (decref(&vn->ref) == 0) {
+		// Remove vnode from vnodes list
+		if (vn->prev != nil) {
+		       vn->prev->next = vn->next;
+		}
+		if (vn->next != nil) {
+			vn->next->prev = vn->prev;
+		}
+		if (mp->vnodes == vn) {
+			mp->vnodes = vn->next;
+		}
 
-vnode *
-ufs_open_ino(MountPoint *mp, ino_t ino) {
-	vnode *vn;
-	int rcode = ffs_vget(mp, ino, LK_SHARED, &vn);
-	if (rcode) {
-		error("cannot get file to open");
+		// Return to free list
+		vn->next = mp->free_vnodes;
+		mp->free_vnodes->prev = vn;
+		mp->free_vnodes = vn;
+		vn->prev = nil;
 	}
-	return vn;
+
+	qunlock(&mp->vnodes_lock);
 }
 
 int
-lookuppath(MountPoint *mp, char *path, vnode **vn)
+countvnodes(vnode* vn)
```

Comment:
If this is called with something locked you might want to say so. If there is no lock of any kind should we worry?

Reply:
It assumes the vnodes have been locked.  I should add a check though.